### DISCUSSION TITLE
Does an edge always end on an OSM node or there's actually no guarantee of that?

### DISCUSSION THREAD

**Discussion by mogita** [valhalla/valhalla#3978] [Q&A] [answered]
Simple as this question is, having gone through the [ConstructEdges](https://github.com/valhalla/valhalla/blob/master/src/mjolnir/graphbuilder.cc#L147) function I kind of thought so but could make sure due to my poor C++ knowledge.

Could there be a chance that an edge ends on an arbitrary point on the OSM way's "LineString" which is not an OSM node? Thanks!

**Comment by kevinkreiser** [comment, marked as answer]
yes it does always end on an osm node. we do not create any new geometry that isn't already there in osm
